File: cmd/agent/keepalive.py
"""Keepalive loop for Beta9 agent.

Sends periodic keepalive updates to transition machine status
from registered → ready and maintain the ready status.
"""

import logging
import threading
import time
from typing import Callable, Optional

import httpx

from .config import AgentConfig
from .metrics import collect_metrics, MachineMetrics

logger = logging.getLogger("beta9-agent")

# Agent version reported in keepalive
AGENT_VERSION = "0.1.0-python"


class KeepaliveLoop:
    """Manages periodic keepalive updates to the gateway."""

    def __init__(self, config: AgentConfig):
        """Initialize keepalive loop.

        Args:
            config: Agent configuration
        """
        self.config = config
        self._stop_event = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._consecutive_failures = 0
        self._max_failures = 3

    def start(self) -> None:
        """Start the keepalive loop in a background thread."""
        if self._thread and self._thread.is_alive():
            logger.warning("Keepalive loop already running")
            return

        self._stop_event.clear()
        self._thread = threading.Thread(
            target=self._run_loop,
            name="keepalive-loop",
            daemon=True,
        )
        self._thread.start()
        logger.info(f"Started keepalive loop (interval: {self.config.keepalive_interval}s)")

    def stop(self) -> None:
        """Stop the keepalive loop."""
        self._stop_event.set()
        if self._thread:
            self._thread.join(timeout=5)
            logger.info("Stopped keepalive loop")

    def _run_loop(self) -> None:
        """Main keepalive loop."""
        # Send first keepalive immediately
        self._send_keepalive()

        while not self._stop_event.is_set():
            # Wait for interval or stop signal
            if self._stop_event.wait(timeout=self.config.keepalive_interval):
                break

            self._send_keepalive()

    def _send_keepalive(self) -> bool:
        """Send a single keepalive update.

        Returns:
            True if successful, False otherwise
        """
        metrics = collect_metrics()

        payload = {
            "machine_id": self.config.machine_id,
            "provider_name": self.config.provider_name,
            "pool_name": self.config.pool_name,
            "agent_version": AGENT_VERSION,
            "metrics": metrics.to_dict(),
        }

        logger.debug(f"Sending keepalive: cpu={metrics.cpu_utilization_pct:.1f}%, mem={metrics.memory_utilization_pct:.1f}%")

        if self.config.dry_run:
            logger.info("Dry run - skipping keepalive")
            return True

        try:
            headers = {
                "Authorization": f"Bearer {self.config.token}",
                "Content-Type": "application/json",
            }

            with httpx.Client(timeout=10) as client:
                response = client.post(
                    self.config.keepalive_url,
                    json=payload,
                    headers=headers,
                )

                if response.status_code == 200:
                    self._consecutive_failures = 0
                    logger.debug(f"Keepalive successful for machine {self.config.machine_id}")
                    return True

                else:
                    self._consecutive_failures += 1
                    logger.warning(
                        f"Keepalive failed: {response.status_code} - {response.text} "
                        f"(failure {self._consecutive_failures}/{self._max_failures})"
                    )
                    return False

        except httpx.ConnectError as e:
            self._consecutive_failures += 1
            logger.warning(f"Keepalive connection failed: {e}")
            return False

        except Exception as e:
            self._consecutive_failures += 1
            logger.error(f"Keepalive error: {e}")
            return False

    @property
    def is_healthy(self) -> bool:
        """Check if keepalive loop is healthy.

        Returns:
            True if recent keepalives succeeded
        """
        return self._consecutive_failures < self._max_failures


def send_single_keepalive(config: AgentConfig) -> bool:
    """Send a single keepalive update (for testing).

    Args:
        config: Agent configuration

    Returns:
        True if successful
    """
    loop = KeepaliveLoop(config)
    return loop._send_keepalive()

File: pkg/agent/keepalive.go
package agent

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sync/atomic"
	"time"

	"github.com/rs/zerolog/log"
)

const AgentVersion = "0.1.0-go"

// KeepalivePayload is the request body for keepalive updates
type KeepalivePayload struct {
	MachineID    string          `json:"machine_id"`
	ProviderName string          `json:"provider_name"`
	PoolName     string          `json:"pool_name"`
	AgentVersion string          `json:"agent_version"`
	Metrics      *MachineMetrics `json:"metrics"`
}

// KeepaliveLoop manages periodic keepalive updates to the gateway
type KeepaliveLoop struct {
	config              *AgentConfig
	metricsCollector    *MetricsCollector
	consecutiveFailures int32
	maxFailures         int32
	stopCh              chan struct{}
	doneCh              chan struct{}
}

// NewKeepaliveLoop creates a new keepalive loop
func NewKeepaliveLoop(config *AgentConfig) *KeepaliveLoop {
	return &KeepaliveLoop{
		config:           config,
		metricsCollector: NewMetricsCollector(),
		maxFailures:      3,
		stopCh:           make(chan struct{}),
		doneCh:           make(chan struct{}),
	}
}

// Start begins the keepalive loop in a goroutine
func (k *KeepaliveLoop) Start(ctx context.Context) {
	go k.run(ctx)
}

func (k *KeepaliveLoop) run(ctx context.Context) {
	defer close(k.doneCh)

	log.Info().
		Dur("interval", k.config.KeepaliveInterval).
		Msg("Started keepalive loop")

	// Send first keepalive immediately
	k.sendKeepalive(ctx)

	ticker := time.NewTicker(k.config.KeepaliveInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			log.Info().Msg("Keepalive loop stopped (context cancelled)")
			return
		case <-k.stopCh:
			log.Info().Msg("Keepalive loop stopped (stop signal)")
			return
		case <-ticker.C:
			k.sendKeepalive(ctx)
		}
	}
}

// Stop signals the keepalive loop to stop and waits for it to finish
func (k *KeepaliveLoop) Stop() {
	close(k.stopCh)
	// Wait for loop to finish with timeout
	select {
	case <-k.doneCh:
	case <-time.After(5 * time.Second):
		log.Warn().Msg("Keepalive loop did not stop within timeout")
	}
}

// IsHealthy returns true if recent keepalives succeeded
func (k *KeepaliveLoop) IsHealthy() bool {
	return atomic.LoadInt32(&k.consecutiveFailures) < k.maxFailures
}

func (k *KeepaliveLoop) sendKeepalive(ctx context.Context) bool {
	metrics, err := k.metricsCollector.Collect()
	if err != nil {
		log.Warn().Err(err).Msg("Failed to collect metrics")
		metrics = &MachineMetrics{}
	}

	payload := &KeepalivePayload{
		MachineID:    k.config.MachineID,
		ProviderName: k.config.ProviderName,
		PoolName:     k.config.PoolName,
		AgentVersion: AgentVersion,
		Metrics:      metrics,
	}

	log.Debug().
		Float64("cpu_pct", metrics.CpuUtilizationPct).
		Float64("mem_pct", metrics.MemoryUtilizationPct).
		Int("free_gpu", metrics.FreeGpuCount).
		Msg("Sending keepalive")

	if k.config.DryRun {
		log.Info().Msg("Dry run - skipping keepalive")
		return true
	}

	body, err := json.Marshal(payload)
	if err != nil {
		log.Error().Err(err).Msg("Failed to marshal keepalive payload")
		return false
	}

	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequestWithContext(ctx, "POST", k.config.KeepaliveURL(), bytes.NewReader(body))
	if err != nil {
		log.Error().Err(err).Msg("Failed to create keepalive request")
		return false
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", k.config.Token))
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		failures := atomic.AddInt32(&k.consecutiveFailures, 1)
		log.Warn().
			Err(err).
			Int32("failure_count", failures).
			Msg("Keepalive connection failed")
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		atomic.StoreInt32(&k.consecutiveFailures, 0)
		log.Debug().
			Str("machine_id", k.config.MachineID).
			Msg("Keepalive successful")
		return true
	}

	failures := atomic.AddInt32(&k.consecutiveFailures, 1)
	log.Warn().
		Int("status", resp.StatusCode).
		Int32("failure_count", failures).
		Int32("max_failures", k.maxFailures).
		Msg("Keepalive failed")
	return false
}

// SendSingleKeepalive sends a single keepalive update (for testing/once mode)
func SendSingleKeepalive(ctx context.Context, config *AgentConfig) bool {
	loop := NewKeepaliveLoop(config)
	return loop.sendKeepalive(ctx)
}
